src/base/emu-i386/kvm.c

```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
 
 static int init_kvm_vcpu(void);
 
+#if !defined(DISABLE_SYSTEM_WA) || !defined(KVM_CAP_IMMEDIATE_EXIT)
+
+/* compat functions for older complex method of immediate exit
+   using a signal that is blocked outside KVM_RUN but not blocked inside it */
+
+#include <pthread.h>
+
+#ifndef KVM_CAP_IMMEDIATE_EXIT
+#define KVM_CAP_IMMEDIATE_EXIT 136
+#define immediate_exit padding1[0]
+#endif
+
+#define KVM_IMMEDIATE_EXIT_SIG (SIGRTMIN + 1)
+static int kvm_cap_immediate_exit = -1;
+
+static void kvm_compat_set_immediate_exit(int set)
+{
+  if (kvm_cap_immediate_exit == -1) { // setup
+    struct sigaction sa;
+    sigset_t set;
+    struct kvm_signal_mask *mask;
+
+    kvm_cap_immediate_exit = ioctl(kvmfd, KVM_CHECK_EXTENSION, KVM_CAP_IMMEDIATE_EXIT) > 0;
+    if (kvm_cap_immediate_exit)
+      return;
+
+    sa.sa_flags = 0;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_handler = SIG_DFL; // never called
+    sigaction(KVM_IMMEDIATE_EXIT_SIG, &sa, NULL);
+
+    /* block KVM_IMMEDIATE_EXIT_SIG in main thread */
+    sigemptyset(&set);
+    sigaddset(&set, KVM_IMMEDIATE_EXIT_SIG);
+    pthread_sigmask(SIG_BLOCK, &set, NULL);
+
+    /* don't block any signals inside the guest */
+    sigemptyset(&set);
+    mask = malloc(sizeof(*mask) + sizeof(set));
+    mask->len = 8;
+    memcpy(mask->sigset, &set, sizeof(set));
+    if (ioctl(vcpufd, KVM_SET_SIGNAL_MASK, mask) == -1) {
+      perror("KVM: KVM_SET_SIGNAL_MASK");
+      leavedos_main(99);
+    }
+    free(mask);
+  }
+
+  if (kvm_cap_immediate_exit)
+    return;
+
+  if (set)
```

Comment:
Does this work?

Reply:
I tested it but the shadowing is confusing for sure, so using sigset now.